Fix 1d cell type nface return value

j8xixo12 opened the pull request:
`CellType` `nface()` only considers 2, 3 dimension, however `CellType` also contains 1 dimension cell type,
`face()` should considering 1 dimension case.

In this PR I also updated a test case to test 1 dimension cell type behavior.

j8xixo12 committed:
add LINE cell type test case

Adding LINE cell type test case to check it's behavior and attributes are expected.

j8xixo12 committed:
fix CellType 1 dim nface return value

CellType nface only considers cases of 2 and 3 dimensions.

However, CellType also contains 1 dimension cell, therefore it should
consider 1 dimension cell as well, or modmesh will allocate incorrect
buffer size and leading to a segementation fault.

j8xixo12 commented on tests/test_mesh.py:
```diff
@@ -201,4 +201,28 @@ def test_3d_single_tetrahedron(self):
+                          nedge=2)
+
+        # _do_metric do nothing due to dim == 1
+        self._check_metric_trivial(mh)
```
I haven't completed the code study in `build_boundary` and `build_ghost`, so I decided not to call them in test case until I complete the code study.

yungyuc replied:
It's OK.  Maybe leave a comment for future work.

yungyuc replied:
Is there a reason not to leave a comment?

yungyuc commented on cpp/modmesh/mesh/StaticMesh.hpp:
```diff
@@ -92,7 +92,15 @@ struct CellType : NumberBase<int32_t, double>
     uint8_t nsurface() const { return m_nsurface; }
 
-    uint8_t nface() const { return 2 == m_ndim ? nedge() : nsurface(); }
+    uint8_t nface() const
```
`nface()` should be a fast function.  It would be great if it has not any branch (if/else), but the original ternary expression already broke it.  When fixing the behavior bug, it would be great to also fix the performance issue.  Could you please take a look?  One way to do it is to compare the generated assembly.

j8xixo12 replied:
I can take a look on this issue!

The technique that first comes to my mind is table lookup, but it needs to maintain a table to store function's informations.

I'm trying to search more informations about branchless code technique.

yungyuc reviewed (changes requested):
This is good start.  I left a comment for the performance concern.  Let's discuss how to deal with it.

j8xixo12 committed:
enhance nface with branchless implementation

Using a unit8_t array to store node, edge, and surface number, then
nface can return the value according to dimensions without branch.

I also update tests/test_mesh.py comments to indicate that there are
some test cases need to be added in the future.

j8xixo12 commented:
I use a array to store `m_nnode`, `m_nedge` and `m_nsurface`, then `nface` can get these information without any branch, just using index to access it.

I also compare switch case version and branchless version assembly code in **debug build**, branchless version has shorter assembly code (27 lines to 24 lines), in release build I can not find `struct CellType` symbol, it should be striped.

But I think this commit may reduced the readability of `struct CellType`.

yungyuc commented on cpp/modmesh/mesh/StaticMesh.hpp:
```diff
@@ -75,10 +75,10 @@ struct CellType : NumberBase<int32_t, double>
-        , m_nedge(nedge_in)
-        , m_nsurface(nsurface_in)
     {
+        m_attrs[NNODE_IDX] = nnode_in;
```
Why not using initialization list like:

```cpp
, m_attrs{0, 0, 0}
```

j8xixo12 replied:
I fixed this in the new commit.

yungyuc commented on cpp/modmesh/mesh/StaticMesh.hpp:
```diff
@@ -112,12 +111,13 @@ struct CellType : NumberBase<int32_t, double>
-    uint8_t m_nnode = 0;
-    uint8_t m_nedge = 0;
-    uint8_t m_nsurface = 0;
+    uint8_t m_attrs[3] = {0};
```
Why `= {0}` rather than `= {0, 0, 0}`?  Invalid syntax?

If it's valid, the latter reads more explicit and less error-prone.

j8xixo12 replied:
`= {0}` and `= {0,0,0}` are valid syntax, but for more readability and less error-prone `= {0,0,0}` would be better.
I will fix it.

yungyuc commented on cpp/modmesh/mesh/StaticMesh.hpp:
```diff
@@ -88,11 +88,10 @@ struct CellType : NumberBase<int32_t, double>
+    uint8_t nnode() const { return m_attrs[NNODE_IDX]; }
+    uint8_t nedge() const { return m_attrs[NEDGE_IDX]; }
+    uint8_t nsurface() const { return m_attrs[NSURFACE_IDX]; }
+    uint8_t nface() const { return m_attrs[m_ndim % 3]; } // Using mod can prevent idx out-of-boundary occurs.
```
Please just use `m_ndim` to avoid the unnecessary mod.  `m_ndim` should never be larger than 3.

j8xixo12 replied:
I use `m_ndim - 1` instead of **mod**.

j8xixo12 commented on tests/test_mesh.py:
```diff
@@ -201,4 +201,30 @@ def test_3d_single_tetrahedron(self):
+
+        # _do_metric do nothing due to dim == 1
+        self._check_metric_trivial(mh)
+        # TODO: Need to add build_boundary and build_ghost to make sure
```
I left a comment to mention that there are some test need to be added in the future.

yungyuc replied:
I see.  Thanks.

j8xixo12 committed:
improve struct CellType code quality
